lib/mayaUsd/fileio/translators/translatorBlendShape.cpp

```diff
@@ -0,0 +1,299 @@
+            _AddBlendShape(
+                blendFn,
+                originalShape,
+                parentTransform,
+                originalPoints,
+                pointIndices,
+                inBetweenName,
+                inBetweenDeltaPoints,
+                targetIdx,
+                ibWeight);
+        }
+    }
+    return true;
+}
+
+bool UsdMayaTranslatorBlendShape::ReadAnimations(
+    const UsdSkelSkinningQuery&     skinningQuery,
+    const UsdSkelAnimQuery&         animQuery,
+    const UsdMayaPrimReaderContext& context)
+{
+    MStatus status;
+
+    const UsdPrim& skinnedPrim = skinningQuery.GetPrim();
+    VtTokenArray   skinnedBlendShapes;
+    skinningQuery.GetBlendShapeOrder(&skinnedBlendShapes);
+    // mesh doesn't have any blendshapes
+    if (skinnedBlendShapes.empty()) {
+        return true;
+    }
+
+    const VtTokenArray animBlendShapes = animQuery.GetBlendShapeOrder();
+    if (animBlendShapes.empty()) {
+        // there are no animated blendShapes. Nothing to do
+        return true;
+    }
+
+    std::vector<unsigned int> correctBlendShapeIndices;
+    correctBlendShapeIndices.reserve(skinnedBlendShapes.size());
+    // The animation prim can hold blendShapes from several mesh prims.
+    // So it's necessary to find the the indices only for those on the current mesh prim.
+    for (const TfToken& token : skinnedBlendShapes) {
+        const auto iter = std::find(animBlendShapes.begin(), animBlendShapes.end(), token);
+        const auto idx = static_cast<int>(std::distance(animBlendShapes.begin(), iter));
+        if (idx >= 0) {
+            correctBlendShapeIndices.emplace_back(idx);
+        }
+    }
+
+    const UsdStageWeakPtr  stage = skinnedPrim.GetStage();
+    const UsdSkelAnimation skelAnimPrim(animQuery.GetPrim());
+    const UsdAttribute     animWeightsAttribute = skelAnimPrim.GetBlendShapeWeightsAttr();
+    const GfInterval       timeCodeInterval(stage->GetStartTimeCode(), stage->GetEndTimeCode());
+    std::vector<double>    animatedTimeSamples;
+    animWeightsAttribute.GetTimeSamplesInInterval(timeCodeInterval, &animatedTimeSamples);
+
+    if (animatedTimeSamples.empty()) {
+        // no animation in time range. Nothing to do.
+        return true;
+    }
+
+    MTimeArray  timeArray;
+    MTime::Unit timeUnit = MTime::uiUnit();
+    double      timeSampleMultiplier = context.GetTimeSampleMultiplier();
+    timeArray.setLength(animatedTimeSamples.size());
+    for (unsigned int ti = 0u; ti < animatedTimeSamples.size(); ++ti) {
+        timeArray.set(MTime(animatedTimeSamples[ti] * timeSampleMultiplier, timeUnit), ti);
+    }
+
+    MObject  transformNode = context.GetMayaNode(skinnedPrim.GetPath(), false);
+    MDagPath shapeDagPath;
+    status = MDagPath::getAPathTo(transformNode, shapeDagPath);
+    shapeDagPath.extendToShape();
+    MObject shapeObject = shapeDagPath.node(&status);
+
+    if (shapeObject.isNull() || !shapeObject.hasFn(MFn::kMesh)) {
+        return false;
```

Comment:
I'm always unsure why in the import and export code, sometimes things return true on errors (or when not finding things), sometimes false.

Reply:
Yea, not sure what would be the idea behavior here. I try to add errors/warnings when necessary